src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Scalar.cs

```diff
@@ -355,6 +355,91 @@ public static bool Equals(T left, T right)
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint ExtractMostSignificantBit(T value)
+        {
+            if (typeof(T) == typeof(byte))
+            {
+                uint bits = (byte)(object)value;
+                return bits >> 7;
+            }
+            else if (typeof(T) == typeof(double))
+            {
+                ulong bits = BitConverter.DoubleToUInt64Bits((double)(object)value);
+                return (uint)(bits >> 63);
+            }
+            else if (typeof(T) == typeof(short))
+            {
+                uint bits = (ushort)(short)(object)value;
+                return bits >> 15;
+            }
+            else if (typeof(T) == typeof(int))
+            {
+                uint bits = (uint)(int)(object)value;
+                return bits >> 31;
+            }
+            else if (typeof(T) == typeof(long))
+            {
+                ulong bits = (ulong)(long)(object)value;
+                return (uint)(bits >> 63);
+            }
+            else if (typeof(T) == typeof(nint))
+            {
+                if (Environment.Is64BitProcess)
```

Comment:
When are we choosing to use
```C#
#if TARGET_64BIT
```
vs
```C#
if (Environment.Is64BitProcess)
```
?

Reply:
The code in here largely uses `Environment` because it wasn't always in corelib and so it continues using it for consistency.

This is fallback logic and isn't perf critical (it basically only runs on `x86 Unix` and `Arm32`) so I don't think its super important to minimize the IL even more.

Reply:
> This is fallback logic and isn't perf critical (it basically only runs on x86 Unix and Arm32) so I don't think its super important to minimize the IL even more.

That leaves me wondering:
1) We're using things like Unsafe.SkipInit in code that's not perf critical?
2) Does this fallback logic get trimmed out on platforms where it's not used?

Reply:
It still gets used in in indirect calls (delegates/fnptrs/reflection) and so we still need an implementation. This just happens to be the easiest way to do it (mostly because the backing fields are `ulong`, not `T`).

These are all over structs and so all irrelevant code paths are trimmed for each scenario. That is, `byte` won't have paths for `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `nint`, or `nuint`. Likewise, `Environment.Is64BitProcess` is also constant folded by both the JIT and AOT, there isn't actual "cost" here (other than an extra branch preserved in IL).

Reply:
> These are all generics over structs and so all irrelevant code paths are trimmed for each scenario

You're talking about the asm.  I'm wondering about the IL and the linker/trimmer.

Reply:
That I don't know. I know the linker has some understanding of `Is64BitProcess` but I'm not sure under what contexts its used vs not.